### PR TITLE
stm32/sdcard: Allow configuring the SDMMC periph used for SD/MMC card.

### PR DESCRIPTION
**Pull request by dpgeorge**
This can now be selected by setting MICROPY_HW_SDCARD_SDMMC, which defaults
to 1, ie SDMMC1.  This commit also renames the SD pin configuration macros
from MICROPY_HW_SDMMC2_xxx to MICROPY_HW_SDCARD_xxx, as well as renaming
MICROPY_HW_SDMMC_BUS_WIDTH to MICROPY_HW_SDCARD_BUS_WIDTH.

### CODE REVIEW

**Comment by dpgeorge** [comment]
@iabdalkader does this work for you?

File: ports/stm32/sdcard.c [hunk 1]
```diff
@@ -187,7 +177,7 @@ STATIC void sdmmc_msp_init(void) {
 
     #if defined(STM32H7)
     // Reset SDMMC
-    #if defined(MICROPY_HW_SDMMC2_CK)
+    #if MICROPY_HW_SDCARD_SDMMC == 2
     __HAL_RCC_SDMMC2_FORCE_RESET();
     __HAL_RCC_SDMMC2_RELEASE_RESET();
```
Comment (iabdalkader) [inline ports/stm32/sdcard.c:182]:
```suggestion
    SDMMC_FORCE_RESET();
    SDMMC_RELEASE_RESET();
```
Need to define macros for these as well... I missed that in my PR.

**Comment by iabdalkader** [comment]
> @iabdalkader does this work for you?

Yes, but we still need another one for sdio.c... I can send a PR for that similar to this PR, after #7182 is merged.

File: ports/stm32/sdcard.c [hunk 2]
```diff
@@ -187,7 +177,7 @@ STATIC void sdmmc_msp_init(void) {
...
     #else
```
Comment (iabdalkader) [inline ports/stm32/sdcard.c:183]:
```suggestion
    SDMMC_FORCE_RESET();
    SDMMC_RELEASE_RESET();
```
I realized that maybe it's better to define these too at the beginning of the file like all the other macros, just a suggestion.

**Comment by dpgeorge** [comment]
> Does this line in make-pins.py need to change too:

Good catch!  I fixed that, see the commit.

File: ports/stm32/sdcard.c (hunk 2 above)
Comment (dpgeorge) [reply ports/stm32/sdcard.c]: Yes, don.

**Comment by iabdalkader** [comment]
Looks good, thanks! I can pull this locally and test it tomorrow if you like or you can just go ahead an merge it.

**Comment by dpgeorge** [comment]
I'll test it and merge if it works.

**Comment by dpgeorge** [comment]
Tested on PYBv1.1 and PYBD-SF6 with a 2GB SD card.  Works.

**Comment by iabdalkader** [comment]
I sent another PR for SDIO based on this one, see #7189

**Comment by vinz-uts** [comment]
Hi, I'm new on micropython (so, sorry if my question is trivial, but I don't find the resources to understand how to do this). I'm using a STM32 NUCELO-L476RG board, I've cloned `micropython` from git repo and I've compiled it as it's write in the 'ports/stm32/readme.md` file, and all works fine (REPL, import modules, etc.). Now I'm trying to enable SDMMC periph that it's wired to an external SD shield (it works when I use it using STM32CubeIDE and C programming). I don't understand how I can configure it, I've tried to add the line `#define MICROPY_HW_ENABLE_SDCARD    (1)` and `#define MICROPY_HW_SDCARD_SDMMC		(1)` in the `mpconfigboard.h` in `boards/NUCLEO_L476RG` folder, but when the board runs, it waits long time (I think it tries to mount the SD Card) and the message `MPY: can't mount SD card` is shown before starting the REPL:
```
MPY: sync filesystems
MPY: soft reboot
MPY: can't mount SD card
MicroPython v1.17-8-gbbbdef4cc-dirty on 2021-09-09; NUCLEO-L476RG with STM32L476RG
Type "help()" for more information.
>>>
```
Could someone help me to understand/ find the steps to correctly configure the SDMMC periph? Thanks for your support...

**Comment by JohnieBraaf** [comment]
The forum would be a better place to ask this, but you need to define the pins as well, like:

`// SD card detect switch
#define MICROPY_HW_SDCARD_DETECT_PIN        (pin_I8)
#define MICROPY_HW_SDCARD_DETECT_PULL       (GPIO_PULLUP)
#define MICROPY_HW_SDCARD_DETECT_PRESENT    (GPIO_PIN_RESET)

#define MICROPY_HW_SDMMC_D0                 (pin_C8)
#define MICROPY_HW_SDMMC_D1                 (pin_C9)
#define MICROPY_HW_SDMMC_D2                 (pin_C10)
#define MICROPY_HW_SDMMC_D3                 (pin_C11)
#define MICROPY_HW_SDMMC_CK                 (pin_C12)
#define MICROPY_HW_SDMMC_CMD                (pin_D2)`

**Comment by vinz-uts** [comment]
Thanks for your fast reply, I've try to ad this lines in the `mpconfigboard.h` (without detect pins because my shield don't use it) but I still have problems (always `MPY: can't mount SD card` before starting REPL and `OSError: [Errno 19] ENODEV
` if I try to mount the SD using `os` module and `pyb.SDCard` class). I try to move this question on the forum to have support. Thanks for your help.

**Comment by dpgeorge** [comment]
You may find this repo useful, which shows how to define custom boards without forking this repo: https://github.com/micropython/micropython-example-boards (but it doesn't include any examples for SD card config).

**Comment by JohnieBraaf** [comment]
@vinz-uts, if you do not have a detect pin, you likely need to manually set the pyb.Pin.board.SD_SW pin to high.

```
>>> import pyb
>>> import os
>>> pyb.SDCard().present()
False
>>> from machine import Pin
>>> sd_sw = Pin(pyb.Pin.board.SD_SW, Pin.OUT)
>>> sd_sw.value()
0
>>> pyb.SDCard().present()
True
>>> sd_sw.value(1)
>>> pyb.SDCard().present()
False
>>> sd_sw.value(0)
>>> pyb.SDCard().present()
True
>>> os.mount(pyb.SDCard(), '/sd')
>>> os.listdir('/sd')
['bmp.py']
```

Taken from [here](https://github.com/koendv/micropython/tree/devel/ports/stm32/boards/DEVEBOX_STM32H750)

**Comment by vinz-uts** [comment]
@JohnieBraafthanks for your reply. I've tried to do what you suggest. When I start to check if SD is present, the result is already True, but the SD isn't monuted.
```
>>> import pyb
>>> import os
>>> pyb.SDCard().present()
True
>>> from machine import Pin
pyb.SDCard().present()
True
>>> sd_sw = Pin(pyb.Pin.board.SD_SW, Pin.OUT)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'board' has no attribute 'SD_SW'
>>> sd_sw = Pin(pyb.Pin.board.SW, Pin.OUT)
>>> sd_sw.value()
0
>>> pyb.SDCard().present()
True
>>> os.mount(pyb.SDCard(),'/sd')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 19] ENODEV
```
The 'SD_SW' pin is not defined in the pyb.Pin.board, is maybe 'SW' for NUCLEO-L476RG board?
Is correct my configuration in 'mpconfigboard.h'? 
```
#define MICROPY_HW_ENABLE_SDCARD    (1)
#define MICROPY_HW_SDCARD_SDMMC		(1)

#define MICROPY_HW_SDMMC_D0                 (pin_C8)
#define MICROPY_HW_SDMMC_D1                 (pin_C9)
#define MICROPY_HW_SDMMC_D2                 (pin_C10)
#define MICROPY_HW_SDMMC_D3                 (pin_C11)
#define MICROPY_HW_SDMMC_CK                 (pin_C12)
#define MICROPY_HW_SDMMC_CMD                (pin_D2)
```
I've not defined the MICROPY_HW_SDCARD_DETECT_PIN, MICROPY_HW_SDCARD_DETECT_PULL, MICROPY_HW_SDCARD_DETECT_PRESENT pins because my shield doesn't use it (I neither use the D1,D2,D3 pins but they are defined, I think this is not a problem).

**Pull request merged**